output/driver_unix.go

```diff
@@ -183,16 +193,96 @@ func (out *output) openAndSetupMixer() error {
 		return out.alsaError("snd_mixer_selem_get_playback_volume_range", err)
 	}
 
-	// set initial volume and verify it actually works
-	mixerVolume := out.volume*(float32(out.mixerMaxVolume-out.mixerMinVolume)) + float32(out.mixerMinVolume)
-	if err := C.snd_mixer_selem_set_playback_volume_all(out.mixerElemHandle, C.long(mixerVolume)); err != 0 {
-		return out.alsaError("snd_mixer_selem_set_playback_volume_all", err)
+	if out.externalVolume {
+
+		// get current volume from the mixer, and set the spotify volume accordingly
+		var volume C.long
+		C.snd_mixer_selem_get_playback_volume(out.mixerElemHandle, C.SND_MIXER_SCHN_MONO, &volume)
+		out.volume = float32(volume-out.mixerMinVolume) / float32(out.mixerMaxVolume-out.mixerMinVolume)
+
+		out.externalVolumeUpdate <- out.volume
+	} else {
+
+		// set initial volume and verify it actually works
+		mixerVolume := out.volume*(float32(out.mixerMaxVolume-out.mixerMinVolume)) + float32(out.mixerMinVolume)
+		if err := C.snd_mixer_selem_set_playback_volume_all(out.mixerElemHandle, C.long(mixerVolume)); err != 0 {
+			return out.alsaError("snd_mixer_selem_set_playback_volume_all", err)
+		}
 	}
 
+	// set callback and initialize private
+	var cb C.snd_mixer_elem_callback_t = (C.snd_mixer_elem_callback_t)(C.alsaMixerCallback)
+	C.snd_mixer_elem_set_callback(out.mixerElemHandle, cb)
+	C.snd_mixer_elem_set_callback_private(out.mixerElemHandle, unsafe.Pointer(&out.volume))
+
+	go out.waitForMixerEvents()
```

Comment:
Shouldn't this be called only if external volume is enabled?

Reply:
I am not yet sure how the externalVolume flag should behave. I think it should just disable any volume manipulation in golibrespot including the mixer things. 

| Mixer | externalVolume | Behaviour |
|--------|--------|--------|
| off | off | golibrespot handles the volume via sample modulation |
| off | on | golibrespot does nothing when a volume event is received and just plays at 100% all the time |
| on | off | golibrespot synchronizes the volume between the alsa mixer and the volume displayed in the spotify client |
| on | on | undecided here, maybe the mixer volume is updated inside of spotify, but changing the volume in spotify doesn't do anything (?) |